### PR TITLE
feat: create change filters definitions

### PR DESCRIPTION
**Pull request by getinnocuous**

### CODE REVIEW

File: changes.proto [hunk 1]
```diff
@@ -266,11 +273,61 @@ message GetAppSummariesResponse {
...
+  Risk.Severity severity = 1;
+  int32 count = 2;
+}
+message StatusWithCount {
+  ChangeStatus status = 1;
+  int32 count = 2;
+}
+
+message ChangeFiltersResponse {
+  repeated StringWithCount tags = 1;
```
Comment (tphoney) [inline changes.proto:303]: what are tags, in this context ?

File: changes.proto [hunk 2]
```diff
@@ -266,11 +273,61 @@ message GetAppSummariesResponse {
...
+  repeated StringWithCount repos = 2;
```
Comment (tphoney) [inline changes.proto:304]: some companies have multiple organisations, does that need to be considered ?

File: changes.proto [hunk 3]
```diff
@@ -266,11 +273,61 @@ message GetAppSummariesResponse {
...
+  repeated SeverityWithCount risks = 3;
+  repeated StringWithCount authors = 4;
+  repeated StatusWithCount statuses = 5;
+  SortOrder sortOrder = 6; // not sure is this needed in the response?
```
Comment (tphoney) [inline changes.proto:308]: is this comment still relevant, it is worth showing what is the default. btw. what is the default ?

**Review by tphoney — COMMENTED** [review]
a few questions

File: changes.proto (hunk 2 above)
Comment (getinnocuous) [reply changes.proto:304]: I wouldn't say so. Lets just use repo name for now

Comment (DavidS-ovm) [reply changes.proto:304]: repo is really a free-form string for whatever users want to put there. I'd expect them to use mostly "org/repo" or the full repo URL in the worst case, e.g. when they have to distinguish between github and gitlab (which would be rare and terrifying)

File: changes.proto [hunk 4]
```diff
@@ -266,11 +273,61 @@ message GetAppSummariesResponse {
...
+  repeated Risk.Severity risks = 3;
+  repeated string authors = 4;
+  repeated ChangeStatus statuses = 5;
+  SortOrder sortOrder = 6;
+}
+
+message StringWithCount {
+  string name = 1;
+  int32 count = 2;
+}
```
Comment (DavidS-ovm) [inline changes.proto:291]:
```suggestion
message NameWithCount {
  string name = 1;
  int32 count = 2;
}
```

be specific.

File: changes.proto [hunk 5]
```diff
@@ -266,11 +273,61 @@ message GetAppSummariesResponse {
 
 message ListHomeChangesRequest {
   PaginationRequest pagination = 1;
+  ChangeFiltersRequest filters = 2;
+}
+
+message ChangeFiltersRequest {
+  repeated string tags = 1;
+  repeated string repos = 2;
```
Comment (DavidS-ovm) [inline changes.proto:281]:
```suggestion
  repeated string repos = 1;
  repeated string tags = 2;
```

most broad filters first.

File: changes.proto [hunk 6]
```diff
@@ -154,6 +154,13 @@ service ChangesService {
   // Gets the full diff of everything that changed as part of this "change".
   // This includes all items and also edges between them
   rpc GetDiff(GetDiffRequest) returns (GetDiffResponse);
+
+  // List all the available repos for all changes
+  rpc ListChangeRepos(ListChangeReposRequest) returns (ListChangeReposResponse);
+  // List all the available authors for all changes
+  rpc ListChangeAuthors(ListChangeAuthorsRequest) returns (ListChangeAuthorsResponse);
+  // List all the available statuses for all changes
+  rpc ListChangeStatuses(ListChangeStatusesRequest) returns (ListChangeStatusesResponse);
```
Comment (DavidS-ovm) [inline changes.proto:163]: Wouldn't you want all this information in a single call?

File: changes.proto [hunk 7]
```diff
@@ -266,11 +273,61 @@ message GetAppSummariesResponse {
...
+  ChangeStatus status = 1;
+  int32 count = 2;
+}
+
+message ChangeFiltersResponse {
+  repeated NameWithCount tags = 1;
+  repeated NameWithCount repos = 2;
+  repeated SeverityWithCount risks = 3;
+  repeated NameWithCount authors = 4;
+  repeated StatusWithCount statuses = 5;
```
Comment (DavidS-ovm) [inline changes.proto:307]: What are the counts here supposed to mean? And why "ChangeFiltersResponse" at all?

**Review by DavidS-ovm — COMMENTED** [review]
:question:

File: changes.proto (hunk 2 above)
Comment (tphoney) [reply changes.proto]: i have had customers do that, cisco comes to mind with their fragmentation. but cool. 

i had to ask

Comment (DavidS-ovm) [reply changes.proto]: why am I not surprised about cisco? :-D

Still best left to customer's choice. We can still add additional validation and parsing later.

File: changes.proto (hunk 7 above)
Comment (getinnocuous) [reply changes.proto:307]: Ignore this "count" field. Just realised we can just count the array length on the frontend.

**Review by DavidS-ovm — APPROVED** [review]
:man_cook: :kiss:

File: changes.proto [hunk 8]
```diff
@@ -266,13 +269,30 @@ message GetAppSummariesResponse {
...
+  ChangeFiltersRequest filters = 2;
+}
+
+message ChangeFiltersRequest {
+  repeated string repos = 1;
+  map<string, string> tags = 2;
+  repeated Risk.Severity risks = 3;
+  repeated string authors = 4;
+  repeated ChangeStatus statuses = 5;
+  SortOrder sortOrder = 6;
```
Comment (dylanratcliffe) [inline changes.proto:281]: Sorry to jump on this after it's been merged @getinnocuous but shouldn't this all be optional? I.e. it's optional to pass the filters in the first place (if you want unfiltered), then also each of the filters within `ChangeFiltersRequest` is optional?

Comment (DavidS-ovm) [reply changes.proto:281]: All of those fields are collections which can be empty, except for `sortOrder` which always has to be chosen.

**Pull request merged**